mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -4798,6 +4798,7 @@ struct FoldTensorCastPackOp : public OpRewritePattern<PackOp> {
     PackOp newOp = rewriter.create<PackOp>(
         op.getLoc(), newOperands[0], newOperands[1], op.getInnerDimsPos(),
         newMixedTileSizes, op.getPaddingValue(), op.getOuterDimsPerm());
+    newOp->setDiscardableAttrs(op->getDiscardableAttrDictionary());
```

Comment:
Why is this sounds? How do you guarantee that the pattern can't invalidate anything encoded in there?

We don't have a principled approach to this in general unfortunately, here the pattern replace the op so it was visible, other folders operate in-place and are propagating "by accident".

By the way: can't this pattern just operate in-place here instead?

Reply:
I'm not sure how easy it would be to make it operate in place because it is potentially updating a list of mixed operands (`newMixedTileSizes`) meaning operand segment lengths need to be updated as well and I've had a bad experience trying to do that in the past.

> Why is this sounds? How do you guarantee that the pattern can't invalidate anything encoded in there?

There is no guarantee this is sound in the same way any pattern that operates on generic tensor encodings isn't sound. As I see it there are hardly any "globally sound" decisions we can make about arbitrary unstructured attributes so instead it seems as though people have been going for a practical middle ground. This is also consistent with every other cast folder I've seen (not saying there aren't counter examples).

Reply:
I dont know if it can be inplace. The result type is changing I think. 

I think the working assumption is that preserving these attributes are, at best, best effort. You cant always transfer over attributes, so its safe to discard. But in cases like this when the op is being replaced by an op of the same type, its also safe to carry it over. If someone relies on this attribute thats "at your risk", but best effort to preserve the attribute. Its not the most principled thing, but more a pragmatic thing.

Reply:
There are many things that are, let's be positive and call them "quirky", about this dialect. If this were its own project, the approach to attributes for things at this layer would be top of the list of things to document. The very next thing I would do is likely remove most of the canonicalizers and sort them into things that could be documented and used in a targeted way. It isn't even clear to me that the classic definitions even apply to things at this layer. Lots of work, though, and in the absence of it, consistency and pragmatism is a higher valued commodity.

Reply:
> I dont know if it can be inplace. The result type is changing I think.

That's not a problem: result types can be changed in place. The `operand_segment_sizes` that @qedawkins mentioned is however more tricky to manipulate: I don't know what kind of helpers we have to mutate operand groups?

> But in cases like this when the op is being replaced by an op of the same type, its also safe to carry it over.

I don't quite follow why the op being the same type is enough. Since @qedawkins mentioned the `operand_segment_size`, the trivial example is an attribute that refers to a particular operand, for example `my.important_info = #some.property<operand #2, ....>`.
After resizing the operand groups, this annotation would refer to the wrong operand.

Reply:
This conversations seemed weirdly resolved, but it wasn't, I don't think the discussion was at a staged where this PR should have been merged.

Reply:
All I saw were comments about things that the core infra may want to consider more strongly at some point, not related to this particular effort. The contributors were pretty clear.

Reply:
I don't think it has to do with "core infra" right now: it has to do with principles of handling Discardable attributes: we need something more principled here. I don't believe I got satisfying answers on this right now, I mentioned two possible paths:

1) Dropping attributes (which was done before)
2) Not performing the transformations when discardable attributes are present.

I'm not sure what kind of attribute interface could allow to perform this kind of transformation safely (haven't seen a proposal), but does not mean it's not possible either (or other strategies).

Reply:
I don't believe there are satisfying answers in the current state, and we're not going to keep discussing misalignments of the core principles with realities of these dialects. 

We value consistency with the rest of the dialect over a solution that no one seems able to articulate. 

Thank you for your feedback.

Reply:
Sorry, but you can't just discard my comments like this. Please consider these as blocking discussions right now.

Reply:
I'm sorry, but I don't see how you have the standing on this project to say that. Your feedback and insights are of course welcome, but the people with the miles in on these dialects need to arbitrate the result.

Reply:
I don't know what qualifies you to make this call right now, I don't recognize you this authority, and nothing in our practices or documented policies on code reviews support your behavior here.

The consistency of MLIR is important, and as long as this code is under `mlir/....`, I consider qualified to make these comments and totally expect to be taken seriously. I also consider myself fully qualify to revert any patch which lands in these conditions.